Fix Clear Amulet visual bug + Tests

AlexOn1ine opened the pull request:
## Description
Fixes Clear Amulet visual bug. A wrong message was displayed when Intimidate on Clear Amulet was used.
Also removed some code in the Intimidate script that was not needed and added a couple of tests.

![pokeemerald-5](https://github.com/rh-hideout/pokeemerald-expansion/assets/93446519/60479d05-4cdc-4555-a6c5-98714ae77347)
![pokeemerald-6](https://github.com/rh-hideout/pokeemerald-expansion/assets/93446519/9f29efb6-32d7-4b6f-a5c9-5b7109cf2359)

##  **Issue(s) that this PR fixes **
Fixes #3054

## **Discord contact info**
rainonline

AsparagusEduardo commented on test/ability_clear_body.c:
```diff
@@ -0,0 +1,27 @@
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
```suggestion
}

TO_DO_BATTLE_TEST("Clear Body prevents stat stage reduction from moves"); // Growl, Leer, Confide, Fake Tears, Scary Face, Sweet Scent, Sand Attack (Attack, Defense, Sp. Attack, Sp. Defense, Speed, Evasion, Accuracy
TO_DO_BATTLE_TEST("Clear Body prevents Sticky Web");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent stat stage reduction from moves used by the user"); // e.g. Superpower
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Speed reduction from Iron Ball");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Speed reduction from paralysis");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Attack reduction from burn");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent receiving negative stat changes from Baton Pass");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Topsy-Turvy");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Spectral Thief from resetting positive stat changes");
TO_DO_BATTLE_TEST("Clear Body is ignored by Mold Breaker");
```

AsparagusEduardo commented on test/ability_white_smoke.c:
```diff
@@ -0,0 +1,27 @@
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
```suggestion
}

TO_DO_BATTLE_TEST("White Smoke prevents stat stage reduction from moves"); // Growl, Leer, Confide, Fake Tears, Scary Face, Sweet Scent, Sand Attack (Attack, Defense, Sp. Attack, Sp. Defense, Speed, Evasion, Accuracy
TO_DO_BATTLE_TEST("White Smoke prevents Sticky Web");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent stat stage reduction from moves used by the user"); // e.g. Superpower
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Speed reduction from Iron Ball");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Speed reduction from paralysis");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Attack reduction from burn");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent receiving negative stat changes from Baton Pass");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Topsy-Turvy");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Spectral Thief from resetting positive stat changes");
TO_DO_BATTLE_TEST("White Smoke is ignored by Mold Breaker");
```

AsparagusEduardo commented on test/ability_full_metal_body.c:
```diff
@@ -0,0 +1,27 @@
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
```suggestion
}

TO_DO_BATTLE_TEST("Full Metal Body prevents stat stage reduction from moves"); // Growl, Leer, Confide, Fake Tears, Scary Face, Sweet Scent, Sand Attack (Attack, Defense, Sp. Attack, Sp. Defense, Speed, Evasion, Accuracy
TO_DO_BATTLE_TEST("Full Metal Body prevents Sticky Web");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent stat stage reduction from moves used by the user"); // e.g. Superpower
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Speed reduction from Iron Ball");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Speed reduction from paralysis");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Attack reduction from burn");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent receiving negative stat changes from Baton Pass");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Topsy-Turvy");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Spectral Thief from resetting positive stat changes");
TO_DO_BATTLE_TEST("Full Metal Body is ignored by Mold Breaker");
```

AsparagusEduardo commented on test/ability_inner_focus.c:
```diff
@@ -0,0 +1,28 @@
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
```suggestion
}

TO_DO_BATTLE_TEST("Inner Focus prevents flinching");
TO_DO_BATTLE_TEST("Inner Focus is ignored by Mold Breaker");
```

AsparagusEduardo commented on test/ability_own_tempo.c:
```diff
@@ -0,0 +1,28 @@
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
```suggestion
}

TO_DO_BATTLE_TEST("Own Tempo prevents confusion from moves by the opponent");
TO_DO_BATTLE_TEST("Own Tempo prevents confusion from moves by the user");
TO_DO_BATTLE_TEST("Own Tempo prevents confusion from items"); // e.g. Iapapa Berry
TO_DO_BATTLE_TEST("Own Tempo cures confusion if it's obtained via Skill Swap");
TO_DO_BATTLE_TEST("Own Tempo cures confusion obtained from an opponent with Mold Breaker");
```

AsparagusEduardo commented on test/ability_scrappy.c:
```diff
@@ -0,0 +1,28 @@
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
```suggestion
}

TO_DO_BATTLE_TEST("Scrappy allows to hit Ghost-type Pokémon with Normal- and Fighting-type moves");
TO_DO_BATTLE_TEST("Scrappy doesn't bypass a Ghost-type's Wonder Guard");
```

AsparagusEduardo commented on test/hold_effect_clear_amulet.c:
```diff
@@ -0,0 +1,57 @@
+}
+
+SINGLE_BATTLE_TEST("Clear Amulet prevents stat reducing effects")
+{
```
To make sure every stat change is handled, let's use `PARAMETRIZE` to add:
- Leer (Defense)
- Confide (Sp. Attack)
- Fake Tears (Sp. Defense)
- Scary Face (Speed)
- Sweet Scent (Evasion)
- Sand Attack (Accuracy)

AsparagusEduardo commented on test/hold_effect_clear_amulet.c:
```diff
@@ -0,0 +1,57 @@
+}
+
+SINGLE_BATTLE_TEST("Clear Amulet prevents secondary effects that reduce stats")
+{
```
To make sure every stat change is handled, let's use `PARAMETRIZE` to add:
- Aurora Beam (Attack)
- Snarl (Sp. Attack)
- Psychic (Sp. Defense)
- Bubble Beam (Speed)
- Mud Slap (Accuracy)

Unfortunately, there's no move for evasion that does this.

AlexOn1ine commented:
Added all the TODO tests, wrote some additional tests for Own Tempo, Inner Focus, Scrappy and added the missing Clear Amulet checks. Also fixed a tiny Triple Kick bug (issue #3054)

PacFire commented:
swag

AsparagusEduardo commented:
Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.

AlexOn1ine commented:
> Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.

Oh will avoid doing it, sorry. I have done that to have a clean commit history but I understand.

mrgriffin commented:
> > Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.
> 
> Oh will avoid doing it, sorry. I have done that to have a clean commit history but I understand.

Something I often do is write a message saying that I intend to rebase and force push to clean up the history once the PR is accepted :)

AlexOn1ine commented:
> > > Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.
> > 
> > 
> > Oh will avoid doing it, sorry. I have done that to have a clean commit history but I understand.
> 
> Something I often do as a PR author is write a message saying that I intend to rebase and force push to clean up the history once the PR is accepted :)

Good suggestion! 👍

AsparagusEduardo commented:
> > > > Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.
> > > 
> > > 
> > > Oh will avoid doing it, sorry. I have done that to have a clean commit history but I understand.
> > 
> > 
> > Something I often do as a PR author is write a message saying that I intend to rebase and force push to clean up the history once the PR is accepted :)
> 
> Good suggestion! 👍

Even then, if it's intended to be a single commit, we can squash it ourselves upon merging thanks to GitHub's Squash and Merge option :)

AsparagusEduardo commented on test/hold_effect_clear_amulet.c:
```diff
@@ -0,0 +1,101 @@
+        TURN { MOVE(player, move); }
+    } SCENE {
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, opponent); }
+        if (move == MOVE_GROWL)
+            MESSAGE("Foe Wobbuffet's Attack was not lowered!");
+        else if (move == MOVE_LEER)
+            MESSAGE("Foe Wobbuffet's Defense was not lowered!");
+        else if (move == MOVE_CONFIDE)
+            MESSAGE("Foe Wobbuffet's Sp. Atk was not lowered!");
+        else if (move == MOVE_FAKE_TEARS)
+            MESSAGE("Foe Wobbuffet's Sp. Def was not lowered!");
+        else if (move == MOVE_SCARY_FACE)
+            MESSAGE("Foe Wobbuffet's Speed was not lowered!");
+        else if (move == MOVE_SWEET_SCENT)
+            MESSAGE("Foe Wobbuffet's evasiveness was not lowered!");
+        else if (move == MOVE_SAND_ATTACK)
+            MESSAGE("Foe Wobbuffet's accuracy was not lowered!");
```
My instinct would be to convert this to a switch, but that would increase the vertical space. Hmm...

AlexOn1ine replied:
True, switches here might make more sense.

AsparagusEduardo commented on test/ability_pastel_veil.c:
```diff
@@ -33,18 +33,22 @@ DOUBLE_BATTLE_TEST("Pastel Veil prevents Poison Sting poison on partner")
         OPPONENT(SPECIES_PONYTA_GALARIAN) { Ability(ABILITY_PASTEL_VEIL); }
     } WHEN {
-        TURN { MOVE(player, MOVE_TOXIC); }
+        TURN { MOVE(player, MOVE_TOXIC);
+               MOVE(opponent, MOVE_TACKLE);
+        }
```
```suggestion
        TURN { MOVE(player, MOVE_TOXIC); MOVE(opponent, MOVE_TACKLE); }
```

AlexOn1ine commented:
Added the Wonder Guard message, changes if to switch and fixed misalignment.
